src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
 	}
 }
 
-int system_argv(char *const argv[])
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
-		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
-		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
+	return system_argv_pipe_fd(-1, -1, lhscmd, -1, -1, rhscmd);
+}
 
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
+int system_argv_pipe_fd(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+                        int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	pid_t monitorpid = fork();
+	if (monitorpid == -1) {
+		LOG(NULL, "Failed to create child process to monitor pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
+		return -1;
+	} else if (monitorpid == 0) {
+		int pipefd[2];
+		if (pipe(pipefd) == -1) {
+			LOG(NULL, "Failed to create a pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
 			assert(0);
 		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
+		pid_t lhspid = system_argv_fd_nowait(-1, pipefd[1], -1 , lhscmd);
+		if (close(pipefd[1]) == -1) {
+			LOG(NULL, "Could not close write end of pipe file descriptor", "%d", pipefd[1]);
 			assert(0);
 		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
+		pid_t rhspid = system_argv_fd_nowait(pipefd[0], -1, -1, rhscmd);
+		if (close(pipefd[0]) == -1) {
+			LOG(NULL, "Could not close read end of pipe file descriptor", "%d", pipefd[0]);
+			assert(0);
 		}
+
```

Comment:
You **must** close both pipefd[0] and pipefd[1] between lines 180 and 181. If you don't then if the RHS for example exits for some reason, but the LHS does not then the LHS will never get a SIGPIPE, but will just block in the write to the pipe and you will get a deadlock. Likewise if the LHS exists for some reason, the RHS will block in the read call as it will not get EOF. This is because the process which is running the code in lines 181 to 183 has the file descriptors open.

Reply:
pipefd[0] and pipefd[1] are being closed on lines 176 and 171, respectively.  Maybe it looks different because I'm checking the exit status of all system calls now, include close?

Reply:
Ah, ok.  My first pass broke chunky functions down into smaller one, but I should probably keep going.